python/ray/log.py

```diff
@@ -0,0 +1,195 @@
+        # LogRecords to the driver.
+        self.plain_handler = logging.StreamHandler()
+        self.plain_handler.level = self.level
+        self.plain_handler.formatter = logging.Formatter(fmt="%(message)s")
+
+    def emit(self, record: logging.LogRecord):
+        """Emit the log message.
+
+        If this is a worker, serialize the record and send it to the driver via TCP.
+        If this is the driver, emit the message using the appropriate console handler.
+
+        Args:
+            record: Log record to be emitted
+        """
+        import ray
+
+        if (
+            hasattr(ray, "_private")
+            and ray._private.worker.global_worker.mode
+            == ray._private.worker.WORKER_MODE
+        ):
+            self.plain_handler.emit(record)
+        else:
+            logging.StreamHandler.emit(self, record)
+
+
+def call_once(func: Callable[[None], None]):
+    """Decorator for a function that should only run once.
+
+    Args:
+        func: Function to wrap.
+
+    Returns:
+        A function that only runs once; other calls generate debug messages and do
+        not run.
+    """
+
+    @functools.wraps(func)
+    def wrapped(force=False):
+        caller = inspect.stack()[-1]
+
+        if not (wrapped.last_caller or force):
+            wrapped.last_caller = caller
+            func()
+        else:
+            logging.getLogger(__name__).debug(
+                "Ray logging has already been configured at "
+                f"{wrapped.last_caller.filename}::{wrapped.last_caller.lineno}."
+                " Skipping reconfiguration requested at "
+                f"{caller.filename}::{caller.lineno}."
+            )
+
+    wrapped.last_caller = None
+    return wrapped
+
+
+@call_once
+def generate_logging_config():
+    """Generate the default Ray logging configuration."""
+
+    formatters = {
+        "plain": {
+            "datefmt": "[%Y-%m-%d %H:%M:%S]",
+            "format": "%(asctime)s %(package)s %(levelname)s %(name)s::%(message)s",
+        },
+    }
+    filters = {"context_filter": {"()": ContextFilter}}
+    handlers = {
+        "null": {"class": "logging.NullHandler"},
+    }
+
+    force_plain = os.environ.get("FORCE_PLAIN_LOGGING", False)
+    if isinstance(force_plain, str) and force_plain.isdecimal():
+        force_plain = int(force_plain)
+
+    handlers["default"] = {
+        "()": PlainRayHandler,
+        "formatter": "plain",
+        "filters": ["context_filter"],
+    }
+
+    loggers = {
+        # Default ray logger; any log message that gets propagated here will be logged
+        # to the console
+        "ray": {
+            "level": "INFO",
+            "handlers": ["default"],
+        },
+        # Special handling for ray.rllib: only warning-level messages passed through
+        # See https://github.com/ray-project/ray/pull/31858 for related PR
+        "ray.rllib": {
+            "level": "WARN",
+        },
+    }
+
+    dictConfig(
```

Comment:
how does this work? what's going to happen if user calls the same thing with version 1?

Reply:
See https://docs.python.org/3/library/logging.config.html#dictionary-schema-details - python's logging system requires us to pass a `version` key, and its value must be `1` :shrug:.